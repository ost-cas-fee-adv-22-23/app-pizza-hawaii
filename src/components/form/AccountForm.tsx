import { Button, Form, FormInput, FormPassword, Grid } from '@smartive-education/pizza-hawaii';
import React, { FC, FormEvent, useEffect, useState } from 'react';

export type TAccountFormData = {
	password: string;
	confirmPassword: string;
	userName: string;
	email: string;
	firstName: string;
	lastName: string;
};

type TAccountFormDataKeys = keyof TAccountFormData;

export type TAccountFormErrors = { [key in keyof TAccountFormData]?: string };

export type TAccountForm = {
	onSubmit: (data: TAccountFormData) => { status: boolean; errors?: TAccountFormErrors };
	onCancel?: () => void;
	user?: TAccountFormData;
	isLoading?: boolean;
};

const emptyState: TAccountFormData = {
	password: '',
	confirmPassword: '',
	userName: '',
	email: '',
	firstName: '',
	lastName: '',
};

export const AccountForm: FC<TAccountForm> = ({ onSubmit, user = emptyState, isLoading }) => {
	const [state, setState] = useState(user);
	const [isUntouched, setIsUntouched] = useState(true);
	const [isValid, setIsValid] = useState(false);
	const [isSubmitting, setIsSubmitting] = useState(false);
	const [errors, setErrors] = useState<TAccountFormErrors>({});

	useEffect(() => {
		setState(user);
	}, [user]);

	useEffect(() => {
		// check if all fields are empty
		const isEmpty = Object.values(state).every((value) => value === '');

		// if there are no validation errors and
		// the form is not empty, set isValid to true
		setIsValid(Object.keys(errors).length === 0 && !isEmpty);
	}, [state, errors]);

	useEffect(() => {
		// check if passwords match
		if (state.password !== state.confirmPassword) {
			setErrors((prev) => ({ ...prev, confirmPassword: 'Die Passwörter stimmen nicht überein.' }));
		} else {
			setErrors((prev) => ({ ...prev, confirmPassword: undefined }));
		}
	}, [state]);

	const validateFields = (fields: HTMLFormControlsCollection) => {
		// get all fields that are not valid (empty if required, mail adress malformed, etc.)
		const invalidFields = Array.from(fields).filter((field) => {
			const { validity } = field as HTMLInputElement;

			return !validity.valid;
		});

		// get validation messages for all invalid fields
		const validationMessages = invalidFields.reduce((acc, field) => {
			const { name, validationMessage } = field as HTMLInputElement;

			acc[name as keyof TAccountFormData] = validationMessage;

			return acc;
		}, {} as TAccountFormErrors);

		return validationMessages;
	};

	const onSubmitHandler = (e: FormEvent) => {
		e.preventDefault();

		setIsSubmitting(true);

		// get all fields from event target
		const fields = (e.target as HTMLFormElement).elements;

		// validate fields
		const validationMessages = validateFields(fields);

		// if it's not ready to submit, set errors and return
		if (Object.keys(validationMessages).length) {
			setErrors(validationMessages);
			setIsSubmitting(false);
			return;
		}

		// if it's ready to submit, call onSubmit function
		const res = onSubmit(state);

		// if the request was successful, reset state otherwise set errors
		if (res.status === true) {
			setState(emptyState);
		} else {
			setErrors(res.errors || {});
		}
		setIsSubmitting(false);
	};

	const onFieldChange = (e: FormEvent): void => {
		const { name, value } = e.target as HTMLInputElement;

		// update state
		setState({
			...state,
			[name]: value,
		});

		// reset error for this field
		setErrors({
			...errors,
			[name]: undefined,
		});

		// check if form is untouched
		const isUntouched = Object.entries(state).every(([key, value]) => value === user[key as TAccountFormDataKeys]);
		setIsUntouched(isUntouched);
	};

	return (
		<Form onSubmit={onSubmitHandler} noValidate>
			<div className={['mt-4', isLoading && 'opacity-50 animate-pulse'].join(' ')}>
				<Grid variant="col" gap="M" marginBelow="M">
					<FormInput
						name="firstName"
						label="First Name"
						type="text"
						value={state['firstName']}
						onChange={onFieldChange}
						errorMessage={errors['firstName']}
						required
						autoComplete="given-name"
					/>
					<FormInput
						name="lastName"
						label="Last Name"
						type="text"
						value={state['lastName']}
						onChange={onFieldChange}
						errorMessage={errors['lastName']}
						required
						autoComplete="family-name"
					/>
					<FormInput
						name="userName"
						label="Username"
						type="text"
						value={state['userName']}
						onChange={onFieldChange}
						errorMessage={errors['userName']}
						required
						icon="mumble"
						autoComplete="off"
					/>
					<FormInput
						name="email"
						label="E-Mail"
						type="email"
						value={state['email']}
						onChange={onFieldChange}
						errorMessage={errors['email']}
						required
						autoComplete="email"
					/>
					<FormPassword
						name="password"
						label="Passwort"
						value={state['password']}
						onChange={onFieldChange}
						errorMessage={errors['password']}
						required
						autoComplete="off"
					/>
					<FormPassword
						name="confirmPassword"
						label="Passwort wiederholen"
						value={state['confirmPassword']}
						onChange={onFieldChange}
						errorMessage={errors['confirmPassword']}
						required
						autoComplete="off"
					/>
				</Grid>
			</div>

			{/*
				We use the isUntouched or isValid state to enable/disable the submit button only visually.
				We don't want to disable the button completely, because then the user wouldn't be able to see the errors.
			*/}
			<div className={['flex-1', isUntouched || isValid ? 'opacity-50' : undefined].join(' ')}>
				<Button size="L" type="submit" colorScheme="gradient" icon="send" disabled={isSubmitting}>
					Speichern
				</Button>
			</div>
		</Form>
	);
};
