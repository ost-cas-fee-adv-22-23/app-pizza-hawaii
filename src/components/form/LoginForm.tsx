import { Button, Form, FormInput, Label } from '@smartive-education/pizza-hawaii';
import React, { FC, FormEvent, useState } from 'react';

export type TLoginFormData = {
	email: string;
};

export type TLoginFormErrors = { [key in keyof TLoginFormData]?: string };

export type TLoginForm = {
	onSubmit: (data: TLoginFormData) => { status: boolean; errors?: TLoginFormErrors };
	data?: TLoginFormData;
	sectionLabel?: string;
	isLoading?: boolean;
};

const emptyState: TLoginFormData = {
	email: '',
};

export const LoginForm: FC<TLoginForm> = ({ onSubmit, data = emptyState, sectionLabel }) => {
	const [state, setState] = useState(data);
	const [isSubmitting, setIsSubmitting] = useState(false);
	const [errors, setErrors] = useState<TLoginFormErrors>({});

	const onSubmitHandler = (e: FormEvent) => {
		e.preventDefault();

		setIsSubmitting(true);

		// submit form
		const res = onSubmit(state);

		// if the request was successful, reset state otherwise set errors
		if (res.status === true) {
			setState(emptyState);
		} else {
			setErrors(res.errors || {});
		}
		setIsSubmitting(false);
	};

	const onFieldChange = (e: FormEvent): void => {
		const { name, value } = e.target as HTMLInputElement;

		// update state
		setState({
			...state,
			[name]: value,
		});

		// reset error for this field
		setErrors({
			...errors,
			[name]: undefined,
		});
	};

	return (
		<Form onSubmit={onSubmitHandler} noValidate>
			<Label as="legend" size="XL">
				{sectionLabel || 'Login'}
			</Label>
			<FormInput
				name="email"
				label="Email"
				type="email"
				value={state['email']}
				onChange={onFieldChange}
				errorMessage={errors['email']}
				required
				autoComplete="email"
			/>

			<Button size="L" type="submit" colorScheme="gradient" icon="mumble" disabled={isSubmitting}>
				Login
			</Button>
		</Form>
	);
};
