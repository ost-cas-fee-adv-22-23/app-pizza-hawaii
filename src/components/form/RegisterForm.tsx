import React, { useState, FormEvent, FC, useEffect } from 'react';
import { Form, FormInput, Button } from '@smartive-education/pizza-hawaii';

export type TRegisterFormData = {
	fullName: string;
	userName: string;
	email: string;
	password: string;
};

type TRegisterFormErrors = { [key in keyof TRegisterFormData]?: string };

type TRegisterForm = {
	onSubmit: (data: TRegisterFormData) => { status: boolean; errors?: TRegisterFormErrors };
};

const emptyState: TRegisterFormData = {
	fullName: '',
	userName: '',
	email: '',
	password: '',
};

export const RegisterForm: FC<TRegisterForm> = ({ onSubmit }) => {
	const [state, setState] = useState(emptyState);
	const [formIsValid, setFormIsValid] = useState(false);
	const [errors, setErrors] = useState<TRegisterFormErrors>({});

	// check if form is valid
	useEffect(() => {
		// check if all fields are empty
		const isEmpty = Object.values(state).every((value) => value === '');

		// if there are no validation errors and
		// the form is not empty, set formIsValid to true
		setFormIsValid(Object.keys(errors).length === 0 && !isEmpty);
	}, [state, errors]);

	const validateFields = (fields: HTMLFormControlsCollection) => {
		// get all fields that are not valid (empty if required, mail adress malformed, etc.)
		const invalidFields = Array.from(fields).filter((field) => {
			const { validity } = field as HTMLInputElement;

			return !validity.valid;
		});

		// get validation messages for all invalid fields
		const validationMessages = invalidFields.reduce((acc, field) => {
			const { name, validationMessage } = field as HTMLInputElement;

			acc[name as keyof TRegisterFormData] = validationMessage;

			return acc;
		}, {} as TRegisterFormErrors);

		return validationMessages;
	};

	const onSubmitHandler = (e: FormEvent) => {
		e.preventDefault();

		// get all fields from event target
		const { target } = e;
		const fields = (target as HTMLFormElement).elements;

		// validate fields
		const validationMessages = validateFields(fields);

		// if it's not ready to submit, set errors and return
		if (!Object.keys(validationMessages).length) {
			setErrors(validationMessages);
			return;
		}

		// if it's ready to submit, call onSubmit function
		const res = onSubmit(state);

		// if the request was successful, reset state otherwise set errors
		if (res.status === true) {
			setState(emptyState);
		} else {
			setErrors(res.errors || {});
		}
	};

	const onFieldChange = (e: FormEvent): void => {
		const { name, value } = e.target as HTMLInputElement;

		// update state
		setState({
			...state,
			[name]: value,
		});

		// reset error for this field
		setErrors({
			...errors,
			[name]: undefined,
		});
	};

	return (
		<Form onSubmit={onSubmitHandler} noValidate>
			<FormInput
				name="fullName"
				label="Vorname und Name"
				type="text"
				onChange={onFieldChange}
				errorMessage={errors['fullName']}
				required
			/>
			<FormInput
				name="userName"
				label="Username"
				type="text"
				onChange={onFieldChange}
				errorMessage={errors['userName']}
				required
			/>
			<FormInput
				name="email"
				label="E-mail"
				type="email"
				onChange={onFieldChange}
				errorMessage={errors['email']}
				required
			/>
			<FormInput
				name="password"
				label="Password"
				type="password"
				icon="eye"
				onChange={onFieldChange}
				errorMessage={errors['password']}
				required
			/>
			<br />

			{/*
				We use the readyToSubmit state to enable/disable the submit button only visually.
				We don't want to disable the button completely, because then the user wouldn't be able to see the errors.
			*/}

			<div className={formIsValid ? 'opacity-100' : 'opacity-50'}>
				<Button size="L" type="submit" colorScheme="gradient" icon="mumble">
					Let&lsquo; Mumble
				</Button>
			</div>
		</Form>
	);
};
