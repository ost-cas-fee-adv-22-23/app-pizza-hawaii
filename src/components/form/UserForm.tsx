import { Button, Form, FormInput, Grid, Label } from '@smartive-education/pizza-hawaii';
import React, { FC, FormEvent, useEffect, useState } from 'react';

import { TZitadelProfile } from '../../types/Zitadel';

export type TUserFormData = TZitadelProfile;

export type TUserFormErrors = { [key in keyof TUserFormData]?: string };

export type TUserForm = {
	onSubmit: (data: TUserFormData) => { status: boolean; errors?: TUserFormErrors };
	user?: TUserFormData;
	sectionlabel?: string;
};

const emptyState: TUserFormData = {
	firstName: '',
	lastName: '',
	displayName: '',
};

export const UserForm: FC<TUserForm> = ({ onSubmit, user = emptyState, sectionlabel }) => {
	const [state, setState] = useState(user);
	const [formIsValid, setFormIsValid] = useState(false);
	const [errors, setErrors] = useState<TUserFormErrors>({});

	useEffect(() => {
		setState(user);
	}, [user]);

	// check if form is valid
	useEffect(() => {
		console.log('state', state);
		// check if all fields are empty
		const isEmpty = Object.values(state).every((value) => value === '');

		// if there are no validation errors and
		// the form is not empty, set formIsValid to true
		setFormIsValid(Object.keys(errors).length === 0 && !isEmpty);
	}, [state, errors]);

	const validateFields = (fields: HTMLFormControlsCollection) => {
		// get all fields that are not valid (empty if required, mail adress malformed, etc.)
		const invalidFields = Array.from(fields).filter((field) => {
			const { validity } = field as HTMLInputElement;

			return !validity.valid;
		});

		// get validation messages for all invalid fields
		const validationMessages = invalidFields.reduce((acc, field) => {
			const { name, validationMessage } = field as HTMLInputElement;

			acc[name as keyof TUserFormData] = validationMessage;

			return acc;
		}, {} as TUserFormErrors);

		return validationMessages;
	};

	const onSubmitHandler = (e: FormEvent) => {
		e.preventDefault();

		// get all fields from event target
		const { target } = e;
		const fields = (target as HTMLFormElement).elements;

		// validate fields
		const validationMessages = validateFields(fields);

		// if it's not ready to submit, set errors and return
		if (Object.keys(validationMessages).length) {
			setErrors(validationMessages);
			return;
		}

		// if it's ready to submit, call onSubmit function
		const res = onSubmit(state);

		// if the request was successful, reset state otherwise set errors
		if (res.status === true) {
			setState(emptyState);
		} else {
			setErrors(res.errors || {});
		}
	};

	const onFieldChange = (e: FormEvent): void => {
		const { name, value } = e.target as HTMLInputElement;

		// update state
		setState({
			...state,
			[name]: value,
		});

		// reset error for this field
		setErrors({
			...errors,
			[name]: undefined,
		});
	};

	return (
		<Form onSubmit={onSubmitHandler} noValidate>
			<fieldset>
				<Label as="legend" size="XL">
					{sectionlabel || 'Deine Daten'}
				</Label>
				<div className="mt-4">
					<Grid variant="col" gap="M" marginBelow="M">
						<FormInput
							name="displayName"
							label="Display Name"
							type="text"
							value={state['displayName']}
							onChange={onFieldChange}
							errorMessage={errors['displayName']}
							required
						/>
						<FormInput
							name="firstName"
							label="Vorname"
							type="text"
							value={state['firstName']}
							onChange={onFieldChange}
							errorMessage={errors['firstName']}
							required
							autoComplete="given-name"
						/>
						<FormInput
							name="lastName"
							label="Name"
							type="text"
							value={state['lastName']}
							onChange={onFieldChange}
							errorMessage={errors['lastName']}
							required
							autoComplete="family-name"
						/>
					</Grid>
				</div>
			</fieldset>

			<br />

			{/*
				We use the readyToSubmit state to enable/disable the submit button only visually.
				We don't want to disable the button completely, because then the user wouldn't be able to see the errors.
			*/}

			<div className={formIsValid ? 'opacity-100' : 'opacity-50'}>
				<Button size="L" type="submit" colorScheme="gradient" icon="mumble">
					Let&lsquo; Mumble
				</Button>
			</div>
		</Form>
	);
};
