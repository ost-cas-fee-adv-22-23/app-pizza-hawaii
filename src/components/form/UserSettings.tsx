import React, { FC, useEffect, useState } from 'react';

import { TUserFormData, UserForm } from './UserForm';

type TUserSettings = {
	setSuccess?: () => void;
};
const BASE_URL = process.env.NEXT_PUBLIC_VERCEL_URL;
const UserSettings: FC<TUserSettings> = ({ setSuccess }) => {
	const [user, setUser] = useState<TUserFormData>();

	const getUser = async () => {
		const res = await fetch(`${BASE_URL}/api/profile`, {
			method: 'GET',
			headers: {
				'Content-Type': 'application/json',
			},
		});

		if (!res.ok) {
			throw new Error('Failed to fetch user info');
		}

		const data = await res.json();

		if (data?.profile) {
			setUser(data.profile);
		}
	};

	const updateUser = async (data: TUserFormData) => {
		try {
			const res = await fetch(`${BASE_URL}/api/profile`, {
				method: 'PUT',
				headers: {
					'Content-Type': 'application/json',
				},
				body: JSON.stringify(data),
			});

			if (!res.ok) {
				throw new Error(`Failed to update user: ${res.statusText}`);
			}
		} catch (error) {
			throw new Error(error as string);
		}
	};

	useEffect(() => {
		getUser();
	}, []);

	const onSubmit = (data: TUserFormData) => {
		// Update user
		(async () => {
			await updateUser(data);
		})();

		// Call setSuccess if it was passed
		setSuccess && setSuccess();

		// Return true to indicate that the form was submitted successfully for a few milliseconds
		return { status: true };
	};

	return <UserForm user={user as TUserFormData} onSubmit={onSubmit} />;
};

export default UserSettings;
