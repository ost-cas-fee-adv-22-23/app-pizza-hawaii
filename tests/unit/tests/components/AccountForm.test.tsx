import '@testing-library/jest-dom';
import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

import { AccountForm, TAccountFormData } from '../../../../src/components/form/AccountForm';
/**
 * Unit tests for the AccountForm component.
 *
 * Basicly we want to ensure the functionality of the form.
 * We test the following on AccountForm component:
 * 1. Render the AccountForm component with a provided user
 * 2. Render the AccountForm component with a empty user
 * 3. Render the AccountForm component with a wrong user
 * 4. Render the AccountForm component with a loading state
 * 5. Render the AccountForm component with a disabled submit button
 * 6. Render the AccountForm component with a enabled submit button
 * 7. Render the AccountForm component with a disabled submit button if the passwords do not match
 * 8. Render the AccountForm component with a enabled submit button if the passwords match
 * 9. Render the AccountForm component with a disabled submit button if the email is not valid
 * 10. Render the AccountForm component with a enabled submit button if the email is valid
 * 11. Render the AccountForm component with a disabled submit button if the username is not valid
 * 12. Render the AccountForm component with a enabled submit button if the username is valid
 * 13. Render the AccountForm component with a disabled submit button if the firstname is not valid
 *
 **/
const propsUser = {
	user: {
		password: 'strongPA$$W0RD',
		confirmPassword: 'strongPA$$W0RD',
		userName: 'Filiks',
		email: 'filiks.adamski@mumble.ch',
		firstName: 'Felix',
		lastName: 'Adam',
	} as TAccountFormData,
	onSubmit: jest.fn(),
	isLoading: false,
};

const propsEmptyUser = {
	user: {
		password: '',
		confirmPassword: '',
		userName: '',
		email: '',
		firstName: '',
		lastName: '',
	} as TAccountFormData,
	onSubmit: jest.fn(),
	isLoading: false,
};

const propsWrongUser = {
	user: {
		...propsUser.user,
		email: 'wrong.email@here',
		confirmPassword: 'strongPA$$--',
	} as TAccountFormData,
	onSubmit: jest.fn(),
	isLoading: false,
};

describe('AccountForm with a provided user', () => {
	afterEach(cleanup);
	// test if the form returns the correct values when submitted
	test('renders AccountForm component with first and last Name field with value', () => {
		const { container } = render(<AccountForm {...propsUser} />);
		expect(container.querySelector('input[name="firstName"]')).toHaveProperty('value', 'Felix');
		expect(container.querySelector('input[name="lastName"]')).toHaveProperty('value', 'Adam');
	});

	// test if the submit button is disabled when the form is loaded. expect the button container to have the class `opacity-50`
	test('renders AccountForm component with disabled submit button when loading. The Visibility of Submitbutton is 50%', () => {
		const { container } = render(<AccountForm {...propsUser} isLoading={true} />);
		const buttonContainer = container.querySelector('.flex-1.opacity-50');
		expect(buttonContainer).toHaveClass('opacity-50');
		expect(container).toMatchSnapshot();
	});

	// test if the form subits empty strings when the form is empty
	test('renders AccountForm with empty fields if no user Data is provided', () => {
		const { container } = render(<AccountForm {...propsEmptyUser} />);
		expect(container.querySelector('input[name="firstName"]')).toHaveProperty('value', '');
		expect(container.querySelector('input[name="lastName"]')).toHaveProperty('value', '');
	});

	// test if the password maches the confirm password
	test('if two different passwords are entered, the input field text is red', () => {
		const { container } = render(<AccountForm {...propsWrongUser} />);
		expect(container.querySelector('input[name="password"]')).toHaveProperty('value', 'strongPA$$W0RD');
		expect(container.querySelector('input[name="confirmPassword"]')).toHaveProperty('value', 'strongPA$$--');
		expect(container.querySelector('input[name="confirmPassword"]')).toHaveClass('text-error-red');
	});

	// the non-null-assertion operator ! is used to make the test working which indicates
	// that the value is actually not null
	test('state of form is updated when user types in input field.', () => {
		const { container } = render(<AccountForm {...propsUser} />);
		const firstNameInput = container.querySelector('input[name="firstName"]');
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		fireEvent.change(firstNameInput!, { target: { value: 'Martin' } });
		expect(firstNameInput).toHaveProperty('value', 'Martin');
	});
});
