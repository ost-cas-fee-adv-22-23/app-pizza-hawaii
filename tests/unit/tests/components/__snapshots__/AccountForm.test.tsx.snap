// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm with a provided user renders AccountForm component with disabled submit button when loading. The Visibility of Submitbutton is 50% 1`] = `
<div>
  <form
    class="flex w-full flex-col mb-8 gap-x-8 gap-y-4 sm:gap-x-4 text-slate-700 dark:text-slate-300"
    novalidate=""
  >
    <div
      class="mt-4 opacity-50 animate-pulse"
    >
      <div
        class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4 mb-8  "
      >
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":r6:"
          >
            First Name
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="given-name"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 "
              id=":r6:"
              name="firstName"
              required=""
              type="text"
              value="Felix"
            />
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":r7:"
          >
            Last Name
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="family-name"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 "
              id=":r7:"
              name="lastName"
              required=""
              type="text"
              value="Adam"
            />
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":r8:"
          >
            Username
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="off"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 pr-10"
              id=":r8:"
              name="userName"
              required=""
              type="text"
              value="Filiks"
            />
            <span
              class="absolute right-4 top-1/2 -translate-y-1/2 flex text-slate-600"
            >
              <span
                class="inline-flex text-[16px] containing-svg:fill-current containing-svg:w-[1em]"
              >
                <svg
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.793 10H6.207a.5.5 0 0 0-.353.854l1.792 1.792a.5.5 0 0 0 .708 0l1.792-1.793A.5.5 0 0 0 9.793 10Z"
                  />
                  <path
                    d="M14.807 5.371A2.962 2.962 0 0 0 16 3a3 3 0 0 0-3-3 2.96 2.96 0 0 0-2.029.806 5.866 5.866 0 0 0-5.942 0A2.96 2.96 0 0 0 3 0a3 3 0 0 0-3 3c0 .974.472 1.827 1.193 2.371C.015 7.867-.64 11.017.935 13.298 2.173 15.09 4.55 16 8 16s5.827-.91 7.065-2.702c1.575-2.28.92-5.43-.258-7.927Zm-1.387 6.79C12.577 13.381 10.754 14 8 14s-4.577-.62-5.42-1.839c-1.05-1.522-.417-3.836.055-5.079C3.085 5.89 4.808 2 8 2s4.914 3.89 5.365 5.082c.472 1.243 1.106 3.557.055 5.08Z"
                  />
                  <path
                    d="M6.67 6.617c-.114-.442-.38-1.05-1.164-1.05-1.081 0-1.987 1.596-1.997 2.306-.01.71.473 1.483 1.07 1.483.598 0 1.123-.535 1.658-1.297.535-.762.546-.999.432-1.442ZM10.494 5.567c-.784 0-1.05.608-1.163 1.05-.114.443-.103.68.432 1.442.535.762 1.06 1.297 1.658 1.297.597 0 1.08-.772 1.07-1.483-.01-.71-.916-2.306-1.997-2.306Z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":r9:"
          >
            E-Mail
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="email"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 "
              id=":r9:"
              name="email"
              required=""
              type="email"
              value="filiks.adamski@mumble.ch"
            />
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":ra:"
          >
            Passwort
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="off"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 pr-10"
              id=":ra:"
              name="password"
              required=""
              type="password"
              value="strongPA$$W0RD"
            />
            <button
              class="absolute right-4 top-1/2 -translate-y-1/2 inline-flex text-slate-600 p-4 -mx-4"
              title="Click to show password"
              type="button"
            >
              <span
                class="inline-flex text-[16px] containing-svg:fill-current containing-svg:w-[1em]"
              >
                <svg
                  fill="none"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.00013 14C12.7071 14 15.7441 8.716 15.8711 8.492C16.0421 8.188 16.0431 7.816 15.8721 7.512C15.7461 7.287 12.7311 2 8.00013 2C3.24513 2 0.251127 7.289 0.126127 7.514C-0.0428725 7.817 -0.0418725 8.186 0.128127 8.489C0.254128 8.713 3.26913 14 8.00013 14ZM8.00013 4C10.8391 4 13.0361 6.835 13.8181 8C13.0341 9.166 10.8371 12 8.00013 12C5.15913 12 2.96213 9.162 2.18113 7.999C2.95813 6.835 5.14613 4 8.00013 4Z"
                    fill="currentColor"
                  />
                  <path
                    d="M8.00013 10C9.1047 10 10.0001 9.10457 10.0001 8C10.0001 6.89543 9.1047 6 8.00013 6C6.89556 6 6.00013 6.89543 6.00013 8C6.00013 9.10457 6.89556 10 8.00013 10Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="sr-only"
              >
                Click to show password
              </span>
            </button>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-slate-800 dark:text-slate-200 font-semibold text-base mb-xxs "
            for=":rb:"
          >
            Passwort wiederholen
          </label>
          <div
            class="relative"
          >
            <input
              autocomplete="off"
              class="w-full text-black text-base rounded-lg p-4 bg-slate-50 hover:bg-slate-100 focus:bg-white border border-slate-300 outline outline-slate-300 outline-0 hover:border-slate-300 hover:outline-slate-300 hover:outline-1 focus:border-purple-600 focus:outline-purple-600 focus:outline-1 disabled:text-slate-400 disabled:bg-slate-50 disabled:border-slate-300 disabled:outline-slate-300 disabled:outline-0 pr-10"
              id=":rb:"
              name="confirmPassword"
              required=""
              type="password"
              value="strongPA$$W0RD"
            />
            <button
              class="absolute right-4 top-1/2 -translate-y-1/2 inline-flex text-slate-600 p-4 -mx-4"
              title="Click to show password"
              type="button"
            >
              <span
                class="inline-flex text-[16px] containing-svg:fill-current containing-svg:w-[1em]"
              >
                <svg
                  fill="none"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.00013 14C12.7071 14 15.7441 8.716 15.8711 8.492C16.0421 8.188 16.0431 7.816 15.8721 7.512C15.7461 7.287 12.7311 2 8.00013 2C3.24513 2 0.251127 7.289 0.126127 7.514C-0.0428725 7.817 -0.0418725 8.186 0.128127 8.489C0.254128 8.713 3.26913 14 8.00013 14ZM8.00013 4C10.8391 4 13.0361 6.835 13.8181 8C13.0341 9.166 10.8371 12 8.00013 12C5.15913 12 2.96213 9.162 2.18113 7.999C2.95813 6.835 5.14613 4 8.00013 4Z"
                    fill="currentColor"
                  />
                  <path
                    d="M8.00013 10C9.1047 10 10.0001 9.10457 10.0001 8C10.0001 6.89543 9.1047 6 8.00013 6C6.89556 6 6.00013 6.89543 6.00013 8C6.00013 9.10457 6.89556 10 8.00013 10Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="sr-only"
              >
                Click to show password
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-1 opacity-50"
    >
      <button
        class="flex items-center justify-center w-full transition-all duration-300 outline-none outline-offset-0 py-4 px-6 rounded-lg gap-y-1 gap-x-3 text-base font-semibold text-white bg-gradient-to-r from-violet-600 via-pink-500 to-violet-700 bg-pos-0 bg-size-200 hover:bg-pos-100 hover:bg-size-300 hover:outline-violet-100 hover:outline-3 focus:bg-pos-100 focus:bg-size-300 focus:outline-violet-200 focus:outline-4 active:bg-pos-100 active:bg-size-300 active:outline-violet-200 active:outline-4 "
        type="submit"
      >
        <span
          class="inline-block font-semibold overflow-hidden text-ellipsis mb-[-0.2em] pb-[0.2em] text-base"
        >
          Speichern
        </span>
        <span
          class="inline-flex text-[16px] containing-svg:fill-current containing-svg:w-[1em]"
        >
          <svg
            viewBox="0 0 17 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.2071 0.292881C15.9341 0.0208806 15.5271 -0.0721194 15.1641 0.0588806L1.16406 5.05888C0.787056 5.19288 0.526056 5.53988 0.502056 5.93888C0.478056 6.33888 0.694056 6.71388 1.05306 6.89388L5.63906 9.18588L11.5001 4.99988L7.31306 10.8619L9.60506 15.4479C9.77606 15.7869 10.1231 15.9999 10.5001 15.9999C10.5211 15.9999 10.5411 15.9989 10.5611 15.9979C10.9611 15.9739 11.3081 15.7139 11.4431 15.3359L16.4431 1.33588C16.5721 0.972881 16.4801 0.565881 16.2071 0.292881Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </form>
</div>
`;
